pkg/threescale/authorizer/authorizer.go

```diff
@@ -0,0 +1,276 @@
+			conf := &systemCache.config
+			conf.refreshInterval = cache.DefaultCacheRefreshInterval
+		}
+
+		if systemCache.config.ttlSeconds == time.Duration(0) {
+			conf := &systemCache.config
+			conf.ttlSeconds = cache.DefaultCacheTTL
+		}
+
+		go func() {
+			ticker := time.NewTicker(systemCache.config.refreshInterval)
+			for {
+				select {
+				case <-ticker.C:
+					systemCache.cache.Refresh()
+				case <-systemCache.config.stop:
+					ticker.Stop()
+					return
+				}
+			}
+		}()
+
+	}
+
+	return &Manager{
+		clientBuilder: builder,
+		systemCache:   *systemCache,
+	}, nil
+}
+
+// NewSystemCache returns a system cache based on the provided interface and config
+func NewSystemCache(cache cache.ConfigurationCache, config SystemCacheConfig) *SystemCache {
+	return &SystemCache{
+		cache:  cache,
+		config: config,
+	}
+}
+
+func NewSystemCacheConfig(refreshRetries int, refreshInterval, ttlSeconds time.Duration, stopRefreshing chan struct{}) *SystemCacheConfig {
+	return &SystemCacheConfig{
+		numRetryFailedRefresh: refreshRetries,
+		refreshInterval:       refreshInterval,
+		ttlSeconds:            ttlSeconds,
+		stop:                  stopRefreshing,
+	}
+}
+
+// GetSystemConfiguration returns the configuration from 3scale system which can be used to fulfill and Auth request
+func (m Manager) GetSystemConfiguration(systemURL string, request SystemRequest) (client.ProxyConfig, error) {
+	var config client.ProxyConfig
+	var err error
+
+	if err = m.validateSystemRequest(request); err != nil {
+		return config, err
+	}
+
+	if m.systemCache.cache != nil {
+		config, err = m.fetchSystemConfigFromCache(systemURL, request)
+
+	} else {
+		config, err = m.fetchSystemConfigRemotely(systemURL, request)
+	}
+
+	if err != nil {
+		return config, fmt.Errorf("cannot get 3scale system config - %s", err.Error())
+	}
+
+	return config, nil
+}
+
+// AuthRep does a Authorize and Report request into 3scale apisonator
+func (m Manager) AuthRep(backendURL string, request BackendRequest) (*BackendResponse, error) {
+	client, err := m.clientBuilder.BuildBackendClient(backendURL)
+	if err != nil {
+		return nil, fmt.Errorf("unable to build required client for 3scale backend - %s", err.Error())
+	}
+
+	req, err := request.ToAPIRequest()
+	if err != nil {
+		return nil, fmt.Errorf("unable to build request to 3scale - %s", err)
+	}
+
+	res, err := client.AuthRep(*req)
+	if err != nil {
+		return nil, fmt.Errorf("error calling AuthRep - %s", err)
+	}
+
+	return &BackendResponse{Authorized: res.Success()}, nil
+}
+
+// validateSystemRequest to avoid wasting compute time on invalid request
+func (m Manager) validateSystemRequest(request SystemRequest) error {
+	if request.Environment == "" || request.ServiceID == "" || request.AccessToken == "" {
+		return fmt.Errorf("invalid arguements provided")
+	}
+	return nil
+}
+
+func (m Manager) generateSystemCacheKey(systemURL, svcID string) string {
+	return fmt.Sprintf("%s_%s", systemURL, svcID)
+}
```

Comment:
These two funcs (at least) are free functions, that is, they don't need to access `m` at all. Can we keep them here private but disassociate from the `m Manager` parameter?

Reply:
right, yep sure